osu.Framework/Bindables/BindableSize.cs

```
@@ -51,10 +51,13 @@ protected override Size ClampValue(Size value, Size minValue, Size maxValue)
 
         protected override int Compare(Size x, Size y)
         {
-            if (x.Width == y.Width && x.Height == y.Height)
-                return 0;
+            if (x.Width < y.Width || x.Height < y.Height)
+                return -1;
```

I don't think the new code is right either. For example, it is both `(0, 1) < (1, 0)` and `(1, 0) < (0, 1)`.

Can't we just use the lexicographic ordering?
```csharp
if (x.Width != y.Width) return x.Width.CompareTo(y.Width);
return x.Height.CompareTo(y.Height);
```

---

hmm, this would allow `MinValue` to be `(0, 1)` and `MaxValue` to be `(1, 0)`, which is not valid.. not sure how to handle these cases, maybe override and throw undefined exception on that case instead?

---

I missed that the comparison is for comparison between min/max values.
As the ordering required for the size clamping is not a total ordering, but only a partial ordering, `Compare` cannot always return one of the `-1, 0, 1` values for each comparison, but it has to return a separate "unordered" case.
Perhaps the return type of `Compare` should just be `bool`, and return `true` if LHS is less than or equals to RHS. If the ordering between LHS and RHS is undefined (unordered), it returns `false`.
For `Size`, it is `Compare(x, y) => x.Width <= y.Width && x.Height <= y.Height`.

---

`BindableSize` should not implement `IComparable`, period. There is no sane way to declare a full ordering on it as stated above. Where is the requirement of implementing it coming from?

---

doesn't implement `IComparable`, simply using the method for comparing `MinValue` and `MaxValue` to avoid non-overlapping ranges.

---

then this abstraction is wrong. declare a bound-checking method that returns a `bool` in `ConstrainedBindable` instead of that `CompareTo()`.

---

hmm, you have a point. as it isn't used anywhere else.